### ISSUE TITLE
Unable to mock method in shared library class

### ISSUE THREAD

**Issue by rockleez** [jenkinsci/JenkinsPipelineUnit#355]
I'm trying to test a shared library which is a class.
However when i try to run the test i get the exception
**No signature of method: foo.httpRequest()is applicable for argument types: (java.util.LinkedHashMap)**

I can initialize the class in the test but i'm unable to mock the httpRequest due to the class being loaded.
I understand why this is happening but not sure on how to get past this issue.

foo.groovy
```groovy
class foo () {
	Boolean testMode
	foo(Boolean testMode) {
		this.testMode = testMode
	}
	
	private String processRequest(String projectKey) {
		def responseInfo = httpRequest acceptType: "APPLICATION_JSON",
				consoleLogResponseBody: true,
				authentication: "test",
				contentType: 'APPLICATION_JSON',
				ignoreSslErrors: true,
				quiet: true,
				responseHandle: 'NONE',
				url: "testUrl",
				validResponseCodes: '200'
		 
		return responseInfo				
	}
	
	String getRequestInfo(String projectKey) {
		return processRequest(projectKey)
	}
}
```

foo_tests.groovy
```groovy
class fooTests extends BasePipelineTest {    

    @Override
    @Before
    void setUp() {
        super.setUp()
		helper.libLoader.preloadLibraryClasses = true
		helper.registerSharedLibrary(library('jiraServiceDesk')
			    .defaultVersion("<notNeeded>")
			    .allowOverride(true)
			    .implicit(false)
			    .targetPath('<notNeeded>')
			    .retriever(projectSource())
			    .build())
		
		helper.registerAllowedMethod("httpRequest", [LinkedHashMap], { LinkedHashMap request -> 
		    println request
			return "abc"
	    })    
		
	}		
	
	@Test
	void testGetRequestInfo() {		
		def bar = new foo(true)
		assertEquals(bar.getRequestInfo("xyz"), "abc")
	}
	
}
```

**Comment by nre-ableton** [comment]
In your tests, try changing the method registration to simply `		helper.registerAllowedMethod("httpRequest", [Map], { Map request ->`.

**Comment by rockleez** [comment]
> In your tests, try changing the method registration to simply `helper.registerAllowedMethod("httpRequest", [Map], { Map request ->`.

same result unfortunatly
No signature of method: foo.httpRequest() is applicable for argument types: (java.util.LinkedHashMap). I suspect it still looking for the foo reference which it cant find

**Comment by nre-ableton** [comment]
>  No signature of method: foo.httpRequest() is applicable for argument types: (java.util.LinkedHashMap). I suspect it still looking for the foo reference which it cant find

Ah, I think I see what is going on. Is `foo.groovy` a singleton (ie, in the `vars` folder), or a proper class in the `src` folder?

If it's the latter, you need to pass the script context to `foo` so that it can access `httpRequest`, which would look something like this:

```groovy
class foo () {
	Object script
	Boolean testMode

	foo(Object script, Boolean testMode) {
		assert script
		this.script = script
		this.testMode = testMode
	}
	
	private String processRequest(String projectKey) {
		def responseInfo = script.httpRequest acceptType: "APPLICATION_JSON",
				consoleLogResponseBody: true,
				authentication: "test",
				contentType: 'APPLICATION_JSON',
				ignoreSslErrors: true,
				quiet: true,
				responseHandle: 'NONE',
				url: "testUrl",
				validResponseCodes: '200'
		 
		return responseInfo				
	}
	
	String getRequestInfo(String projectKey) {
		return processRequest(projectKey)
	}
}
```

Note that when you construct `foo`, you'll need to pass the script context to it, which is usually `this` from a singleton in the `vars` folder.

**Comment by gene-pavlovsky** [comment]
I've had the same issue, and indeed it's about the script context.
I just followed the `HardMath` example from here to make it all work: https://github.com/jenkinsci/JenkinsPipelineUnit/blob/master/README.md#example-pipeline-library-organization

You really need this `test/resources/EmptyPipeline.groovy` file!

I skipped the `vars/hardmath.groovy` file, as I wanted to test just the `HardMath` class (obviously my real classes are named differently).
